src/core/qgsfield.h

```diff
@@ -208,6 +233,65 @@ class CORE_EXPORT QgsField
      */
     void setDefaultValueExpression( const QString& expression );
 
+    /**
+     * Returns any constraints which are present for the field.
+     * @note added in QGIS 3.0
+     * @see setConstraints()
+     * @see constraintOrigin()
+     */
+    Constraints constraints() const;
+
+    /**
+     * Returns the origin of a field constraint, or ConstraintOriginNotSet if the constraint
+     * is not present on this field.
+     * @note added in QGIS 3.0
+     * @see constraints()
+     */
+    ConstraintOrigin constraintOrigin( Constraint constraint ) const;
+
+    /**
+     * Sets a constraint on the field.
+     * @note added in QGIS 3.0
+     * @see constraints()
+     * @see removeConstraint()
+     */
+    void setConstraint( Constraint constraint, ConstraintOrigin origin = ConstraintOriginLayer );
```

Comment:
I wonder if that should be public API like that or of we want to make that private and have a public

    void setConstraint( Constraint constraint );

instead.
Then make QgsVectorDataProvider a friend class so it can still set the private constraints.